File: onshape_cli/api/featurestudio.py
"""Feature Studio support -- create and edit custom FeatureScript elements.

A Feature Studio holds FeatureScript source that defines custom features. Once
imported into a Part Studio, those custom features can be added like any other
feature via the part-studio add-feature endpoint.
"""

from typing import Any, Dict

from .client import OnshapeClient


class FeatureStudioManager:
    """Manager for Onshape Feature Studios."""

    def __init__(self, client: OnshapeClient):
        self.client = client

    async def create(
        self, document_id: str, workspace_id: str, name: str
    ) -> Dict[str, Any]:
        """Create a new Feature Studio tab."""
        path = f"/api/v6/featurestudios/d/{document_id}/w/{workspace_id}"
        return await self.client.post(path, data={"name": name})

    async def get_contents(
        self, document_id: str, workspace_id: str, element_id: str
    ) -> Dict[str, Any]:
        """Get the FeatureScript source text of a Feature Studio."""
        path = f"/api/v6/featurestudios/d/{document_id}/w/{workspace_id}/e/{element_id}"
        return await self.client.get(path)

    async def set_contents(
        self, document_id: str, workspace_id: str, element_id: str, contents: str
    ) -> Dict[str, Any]:
        """Replace the FeatureScript source text of a Feature Studio."""
        path = f"/api/v6/featurestudios/d/{document_id}/w/{workspace_id}/e/{element_id}"
        return await self.client.post(path, data={"contents": contents})

    async def get_specs(
        self, document_id: str, workspace_id: str, element_id: str
    ) -> Dict[str, Any]:
        """Get the feature specs (custom feature definitions) a Feature Studio exports."""
        path = (
            f"/api/v6/featurestudios/d/{document_id}/w/{workspace_id}/e/{element_id}"
            "/featurespecs"
        )
        return await self.client.get(path)

File: node/src/api/featurestudio.ts
import { readFileSync } from "node:fs";

import type { OnshapeClient } from "./client";

export class FeatureStudioManager {
  constructor(private readonly client: OnshapeClient) {}

  async create(documentId: string, workspaceId: string, name: string): Promise<unknown> {
    return this.client.post(`/api/v6/featurestudios/d/${documentId}/w/${workspaceId}`, { name });
  }

  async getContents(documentId: string, workspaceId: string, elementId: string): Promise<unknown> {
    return this.client.get(`/api/v6/featurestudios/d/${documentId}/w/${workspaceId}/e/${elementId}`);
  }

  async setContents(documentId: string, workspaceId: string, elementId: string, contents: string): Promise<unknown> {
    return this.client.post(`/api/v6/featurestudios/d/${documentId}/w/${workspaceId}/e/${elementId}`, { contents });
  }

  async getSpecs(documentId: string, workspaceId: string, elementId: string): Promise<unknown> {
    return this.client.get(`/api/v6/featurestudios/d/${documentId}/w/${workspaceId}/e/${elementId}/featurespecs`);
  }
}

export function loadText(inline?: string, file?: string): string {
  const raw = file ? readFileSync(file, "utf8") : inline;
  if (!raw) throw new Error("Expected text via --contents or --contents-file");
  return raw;
}
